ray_lightning/ray_ddp.py

```diff
@@ -166,65 +150,88 @@ def train(self):
 
         self.global_to_local = self.get_local_ranks()
 
-        trainer = self.trainer
-        assert trainer is not None
-        trainer_ref = ray.put(trainer)
-        # Don't pickle self.trainer when training remotely.
-        self.trainer = None
+        model = self._model
+        model_ref = ray.put(model)
+        # Don't pickle the model when training remotely.
+        self._model = None
 
         queue = None
         if TUNE_INSTALLED and is_session_enabled():
             # Create communication queue and send to all the workers.
             queue = Queue(actor_options={"num_cpus": 0})
 
         futures = [
-            self.workers[i].execute.remote(self.train_remote, trainer_ref, i,
+            self.workers[i].execute.remote(self.train_remote, model_ref, i,
                                            queue)
             for i in range(self.num_workers)
         ]
 
         results = process_results(futures, queue)
         results, best_path, state_dict = results[0]
-        self.trainer = trainer
-        self.trainer.model.load_state_dict(state_dict)
-        if self.trainer.checkpoint_callback:
-            self.trainer.checkpoint_callback.best_model_path = best_path
+        self._results = results
+        self._model = model
+        self._model.load_state_dict(state_dict)
+        if self.lightning_module.trainer.checkpoint_callback:
+            self.lightning_module.trainer.checkpoint_callback\
+                .best_model_path = best_path
 
         if queue:
             # Shutdown the queue.
             queue.shutdown()
 
         return results
 
+    def post_dispatch(self):
+        """Shutdown the DDP process group and all the Ray actors. """
+
+        def shutdown_remote():
+            torch.distributed.destroy_process_group()
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
+
+        ray.get([w.execute.remote(shutdown_remote) for w in self.workers])
+        for w in self.workers:
+            ray.kill(w, no_restart=True)
+            del w
+        self.workers = []
+
     # All methods below are only executed in remote Ray workers.
 
     def train_remote(self,
-                     trainer: Trainer,
+                     model: LightningModule,
                      global_rank: int,
                      queue: Queue = None):
         """Training function to be executed on each remote worker."""
-        assert isinstance(self, RayAccelerator)
+        assert isinstance(self, RayPlugin)
         # This method should be executed remotely in each worker.
-        self.trainer = trainer
-        self.trainer.accelerator_backend = self
+        self._model = model
+        self.lightning_module.trainer.accelerator_connector\
+            ._training_type_plugin = self
+        self.lightning_module.trainer.accelerator.training_type_plugin = self
```

Comment:
wow that's nasty

Reply:
yeah this is very hacky...but I had to set this back to self otherwise when we unserialize the model, the plugin was referring to a different object.

Reply:
would it be reasonable to propose an abstraction to PTL?

Reply:
yes possibly, though this is specific to the fact that we are serializing the plugin. Let me think about this some more, but I think this should be fine to keep for now.

Reply:
maybe you can override the serialization codepath (like __reduce__ or something) and have it reference a global variable?

Reply:
Though you're right- this wouldn't be a problem in the first place if PTL didn't have this circular reference.